src/formatting/types.rs

```diff
@@ -1036,10 +1036,45 @@ fn join_bounds_inner(
         },
     )?;
 
-    if !force_newline
-        && items.len() > 1
-        && (result.0.contains('\n') || result.0.len() > shape.width)
-    {
+    // Whether retry the function with forced newline is neded:
+    // - Only if result is not already multiline and did not exceed line width;
+    // - and of type `Trait`;
+    // - and any of the internal arrays contains more than one item;
+    let retry_with_force_newline =
+        if force_newline || (!result.0.contains('\n') && result.0.len() <= shape.width) {
+            false
+        } else {
+            if items.len() > 1 {
+                true
+            } else {
+                match items[0] {
+                    ast::GenericBound::Trait(ref poly_trait_ref, ..) => {
+                        let segments = &poly_trait_ref.trait_ref.path.segments;
+                        if segments.len() > 1 {
+                            true
+                        } else {
+                            if let Some(args_in) = &segments[0].args {
+                                match &**args_in {
+                                    ast::AngleBracketed(args) => {
+                                        if args.args.len() > 1 {
+                                            true
+                                        } else {
+                                            false
+                                        }
+                                    }
+                                    _ => false,
+                                }
+                            } else {
+                                false
+                            }
+                        }
+                    }
+                    _ => false,
+                }
+            }
+        };
```

Comment:
:eyes: this is a pretty gnarly looking conditional.

before thinking about the approach, i'd really like to see if this can be refactored (e.g. with different control flows) to reduce the excessive nesting or otherwise extract it out entirely as a separate function to be able to reduce the levels and rightward drift

Reply:
> ... or otherwise extract it out entirely as a separate function to be able to reduce the levels and rightward drift

Refactored by add `is_segment_with_multi_items_array` function which took out part of the conditions nesting.  Also re-based, so old version is not available.